ipwb/replay.py

```diff
@@ -179,6 +180,111 @@ def showMementosForURIRs_sansJS():
     return redirect('/memento/*/' + urir, code=301)
 
 
+def bin_search(iter, key, datetime=None):
+    # Read line encompassing current position
+    ln = iter.readline()
+
+    while ipwbUtils.isCDXJMetadataRecord(ln):
+        ln = iter.readline()
+
+    surtk, datetimeK, rest = ln.split(maxsplit=2)
+    datetimeK = datetimeK.decode()
+
+    matchDegree = getMatchDegree(surt, datetime, surtk, datetimeK)
+    if matchDegree == MementoMatch.EXACTMATCH:
+        return [ln]
+
+    # If further searching required...
+    left = 0
+    # Go to end of seek stream
+    iter.seek(0, 2)
+    right = iter.tell()  # Current position
+
+    lines = set()  # Prevents dupes
+    while (right - left > 1):
+        mid = (right + left) // 2
+        iter.seek(mid)
+        iter.readline()  # Purge rest of current line
+        line = iter.readline()  # Read the next full line
+
+        if len(line) == 0:
+            right = mid
+            continue
+
+        surtk, datetimeK, rest = line.split(maxsplit=2)
+        datetimeK = datetimeK.decode()
+
+        surtk = surtk.rstrip(b"/")
+        key = key.rstrip(b"/")
+
+        matchDegree = getMatchDegree(key, datetime, surtk, datetimeK)
+
+        if matchDegree == MementoMatch.RIGHTKEYWRONGDATE:
+            lines.add(line)
+            # Iterate further to get lines after selection point
+            nextLine = iter.readline()
+            while nextLine:
+                surtk, datetimeK, rest = nextLine.split(maxsplit=2)
+                surtk = surtk.rstrip(b"/")
+                datetimeK = datetimeK.decode()
+
+                matchDegree = getMatchDegree(key, datetime, surtk, datetimeK)
+                if matchDegree == MementoMatch.RIGHTKEYWRONGDATE:
+                    lines.add(nextLine)
+                elif matchDegree == MementoMatch.EXACTMATCH:
+                    # Exact match found while iterating
+                    lines.clear()
+                    lines.add(nextLine)
+                    break
+
+                nextLine = iter.readline()
+
+            # Continue searching until find first instance
+            right = mid
+        elif matchDegree == MementoMatch.EXACTMATCH:
+            lines.clear()
+            lines.add(line)
+
+            break
```

Comment:
While this one will break the outer loop, you might to use return here as well for similarity sake.

```suggestion
        elif matchDegree == MementoMatch.EXACTMATCH:
            return [line]
```

Reply:
Applied in 3bc6e37.